File: src/cliffordplust/rings/cpp/Zsqrt2.cpp
/* Copyright 2024-2025 Olivier Romain, Francis Blais, Vincent Girouard, Marius Trudeau
 * 
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 * 
 *        http://www.apache.org/licenses/LICENSE-2.0
 * 
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

#include <string>
#include <stdexcept>
#include <iostream>
#include <cmath>
#include <boost/multiprecision/cpp_int.hpp> // Include the Boost Multiprecision library header

#include "Rings.hpp"

using namespace boost::multiprecision; // Use the Boost Multiprecision namespace


template <typename T>
Zsqrt2<T>::Zsqrt2(T p, T q) : _p(p), _q(q) {}


template <typename T>
T Zsqrt2<T>::p() const {return _p;}

template <typename T>
T Zsqrt2<T>::q() const {return _q;}


template <typename T>
Zsqrt2<T> Zsqrt2<T>::sqrt2_conjugate() const {return Zsqrt2<T>(_p, -_q);}


template <typename T>
bool Zsqrt2<T>::is_int() const {return _q == 0;}


template <typename T>
Domega<T> Zsqrt2<T>::to_Domega() const {return Domega<T>(-_q, 0, 0, 0, _q, 0, _p, 0);}

template <typename T>
Zomega<T> Zsqrt2<T>::to_Zomega() const {return Zomega<T>(-_q, 0, _q, _p);}

template <typename T>
Dsqrt2 Zsqrt2<T>::to_Dsqrt2() const {return Dsqrt2(_p, 0, _q, 0);}

template <typename T>
D Zsqrt2<T>::to_D() const {
    if (! is_int()) {
        throw std::runtime_error("The number to convert is not an integer. Got " + to_string());
    }

    return D(_p, 0);
}

template <typename T>
T Zsqrt2<T>::to_int() const {
    if (! is_int()) {
        throw std::runtime_error("The number to convert is not an integer. Got " + to_string());
    }
    
    return _p;
}

template <typename T>
float Zsqrt2<T>::to_float() const {
    return static_cast<float>(_p) + static_cast<float>(_q) * static_cast<float>(std::sqrt(2.0));
}

template <typename T>
long double Zsqrt2<T>::to_long_double() const {
    return static_cast<long double>(_p) + static_cast<long double>(_q) * std::sqrt(static_cast<long double>(2.0));
}


template <typename T>
bool Zsqrt2<T>::operator==(const Zsqrt2& other) const {return (_p == other._p) and (_q == other._q);}

template <typename T>
bool Zsqrt2<T>::operator==(const T& other) const {return (_p == other) and (_q == 0);}

template <typename T>
bool Zsqrt2<T>::operator!=(const Zsqrt2& other) const {return !(*this == other);}

template <typename T>
bool Zsqrt2<T>::operator!=(const T& other) const {return !(*this == other);}

template <typename T>
bool Zsqrt2<T>::operator||(const Zsqrt2& other) const {
    Zsqrt2<T> r1(0, 0);
    Zsqrt2<T> r2(0, 0);
    std::tie(std::ignore, r1) = euclidean_div(*this, other);
    std::tie(std::ignore, r2) = euclidean_div(other, *this);
    return (r1 == 0) and (r2 == 0);
}

template <typename T>
Zsqrt2<T> Zsqrt2<T>::operator+(const Zsqrt2& other) const {return Zsqrt2<T>(_p + other._p, _q + other._q);}

template <typename T>
Zsqrt2<T> Zsqrt2<T>::operator-() const {return Zsqrt2<T>(-_p, -_q);}

template <typename T>
Zsqrt2<T> Zsqrt2<T>::operator-(const Zsqrt2& other) const {return *this + (-other);}

template <typename T>
Zsqrt2<T> Zsqrt2<T>::operator*(const Zsqrt2& other) const {
    return Zsqrt2<T>(
        (_p * other._p) + (2 * _q * other._q),
        (_p * other._q) + (_q * other._p)
    );
}


template <typename T>
Zsqrt2<T> Zsqrt2<T>::pow(unsigned int n) const {
    Zsqrt2<T> nth_power = *this;
    Zsqrt2<T> result(1, 0);

    while (n) {
        if (n & 1) {result = result * nth_power;}
        nth_power = nth_power * nth_power;
        n >>= 1;
    }

    return result;
}


template <typename T>
void Zsqrt2<T>::unit_reduce() {
    Zsqrt2<T> n1 = *this;
    Zsqrt2<T> n2(0, 0);

    // If the sign of p and q is different, reduce the number by multiplying by lambda = 1 + sqrt(2)
    if (std::signbit(n1.p()) xor std::signbit(n1.q())) {
        do {
            n1 = n1 * Zsqrt2<T>(1, 1);
        } while (std::signbit(n1.p()) xor std::signbit(n1.q()));

        // Recover the number with the opposite sign
        n2 = n1 * Zsqrt2<T>(-1, 1);

    // If not, reduce the number by multiplying by lambda**-1 = -1 + sqrt(2)
    } else {
        do {
            n1 = n1 * Zsqrt2<T>(-1, 1);
        } while (std::signbit(n1.p()) == std::signbit(n1.q()));
        
        // Recover the number with the opposite sign
        n2 = n1 * Zsqrt2<T>(1, 1);
    }

    // Return the best number, the one with the smallest coefficients p and q
    T merit1 = std::llabs(n1.p()) + std::llabs(n1.q());
    T merit2 = std::llabs(n2.p()) + std::llabs(n2.q());
    if (merit1 < merit2) {
        *this = n1;
    } else {
        *this = n2;
    }
}


template <typename T>
std::string Zsqrt2<T>::to_string() const {
    return std::to_string(_p) + " + " + std::to_string(_q) + "\u221A2";
}

template <typename T>
void Zsqrt2<T>::print() const {std::cout << to_string() << std::endl;}


/// Functions in the Z[\u221A2] ring
template <typename T>
std::tuple<Zsqrt2, Zsqrt2> euclidean_div(const Zsqrt2& num, const Zsqrt2& div) {
    Zsqrt2<T> num_ = num * div.sqrt2_conjugate();
    T den_ = (div * div.sqrt2_conjugate()).p();

    T a = static_cast<T>(std::round(static_cast<float>(num_.p()) / static_cast<float>(den_)));
    T b = static_cast<T>(std::round(static_cast<float>(num_.q()) / static_cast<float>(den_)));

    Zsqrt2<T> q(a, b);
    Zsqrt2<T> r = num - q * div;

    return {q, r};
}

File: src/cliffordplust/Rings/z_sqrt2.py
# Copyright 2024-2025 Olivier Romain, Francis Blais, Vincent Girouard, Marius Trudeau
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.

"""
This module contains the definition of the Zsqrt2 class, which allows for symbolic calculation with elements of 
the ring of quadratic integers with radicand 2 \u2124[\u221A2]. The ring elements have the form a + b\u221A2, 
where a and b are integers. This class is useful among others to solve 1 dimensional grid problems, where 
solutions are found inside this ring. For more information see 
Neil J. Ross and Peter Selinger, Optimal ancilla-free Clifford+T approximation of z-rotations, 
https://arxiv.org/pdf/1403.2975.
"""

from __future__ import annotations

import math
import numbers as num
from decimal import Decimal, getcontext
from typing import Any, Optional


class Zsqrt2:
    """A simple class to do symbolic computation with elements of the ring \u2124[\u221A2].

    The ring element has the form a + b\u221A2, where a and b are integers.

    Attributes:
        a (int): Integer coefficient of the ring element.
        b (int): \u221A2 coefficient of the ring element.
    """

    def __init__(self, a: int, b: int) -> None:
        """Initialize the ring element.

        Args:
            a (int): Integer coefficient of the ring element.
            b (int): \u221A2 coefficient of the ring element.

        Raises:
            TypeError: If a or b are not integers.
        """
        for input in (a, b):
            if not isinstance(input, num.Integral):
                raise TypeError(
                    f"Expected inputs to be of type int, but got {type(input).__name__}."
                )
        self._a: int = a
        self._b: int = b

    @property
    def a(self) -> int:
        """Integer coefficient of the ring element."""
        return self._a

    @property
    def b(self) -> int:
        """\u221A2 coefficient of the ring element."""
        return self._b

    def conjugate(self) -> Zsqrt2:
        """Define the \u221A2-conjugation operation.

        Returns:
            Zsqrt2: \u221A2-conjugate of the ring element.
        """
        return Zsqrt2(self.a, -self.b)

    def __float__(self) -> float:
        """Define the float representation of the ring element."""
        bsqrt = self.b * math.sqrt(2)
        if math.isclose(self.a, -bsqrt, rel_tol=1e-4):
            getcontext().prec = 50
            return float(self.a + self.b * Decimal(2).sqrt())
        return self.a + bsqrt

    def __getitem__(self, i: int) -> int:
        """Access the values of a and b with their index."""
        return (self.a, self.b)[i]

    def __repr__(self) -> str:
        """Define the string representation of the ring element."""
        repr: str = ""
        repr += str(self.a)
        if self.b < 0:
            repr += str(self.b) + "\u221A2"
        else:
            repr += "+" + str(self.b) + "\u221A2"
        return repr

    def __eq__(self, nb: Any) -> bool:
        """Define the equality of Zsqrt2 class."""
        if isinstance(nb, Zsqrt2):
            return self.a == nb.a and self.b == nb.b
        elif isinstance(nb, num.Integral):
            return self.a == nb and self.b == 0
        return False

    def __neg__(self) -> Zsqrt2:
        """Define the negation of a ring element."""
        return Zsqrt2(-self.a, -self.b)

    def __add__(self, nb: Zsqrt2 | int) -> Zsqrt2:
        """Define the summation operation for the Zsqrt2 class.

        Allow summation with integers or Zsqrt2 objects.
        """
        if isinstance(nb, Zsqrt2):
            return Zsqrt2(self.a + nb.a, self.b + nb.b)
        elif isinstance(nb, num.Integral):
            return Zsqrt2(self.a + nb, self.b)
        raise TypeError(f"Summation operation is not defined with {type(nb).__name__}.")

    def __radd__(self, nb: int) -> Zsqrt2:
        """Define the right summation of int with the Zsqrt2 class."""
        return self.__add__(nb)

    def __iadd__(self, nb: Zsqrt2 | int) -> Zsqrt2:
        """Define the in-place summation operation for the class."""
        return self.__add__(nb)

    def __sub__(self, nb: Zsqrt2 | int) -> Zsqrt2:
        """Define the subtraction operation for the Zsqrt2 class.

        Allow subtraction with integers and Zsqrt2 objects.
        """
        if isinstance(nb, Zsqrt2):
            return Zsqrt2(self.a - nb.a, self.b - nb.b)
        elif isinstance(nb, num.Integral):
            return Zsqrt2(self.a - nb, self.b)
        raise TypeError(f"Subtraction operation is not defined with {type(nb).__name__}.")

    def __rsub__(self, nb: int) -> Zsqrt2:
        """Define the right subtraction of int with the Zsqrt2 class."""
        return (-self).__add__(nb)

    def __isub__(self, nb: Zsqrt2 | int) -> Zsqrt2:
        """Define in-place subtraction operation for the class."""
        return self.__sub__(nb)

    def __mul__(self, nb: Zsqrt2 | int) -> Zsqrt2:
        """Define the multiplication operation for the Zsqrt2 class.

        Allow multiplication with integers and Zsqrt2 objects.
        """
        if isinstance(nb, Zsqrt2):
            return Zsqrt2(self.a * nb.a + 2 * self.b * nb.b, self.a * nb.b + self.b * nb.a)
        elif isinstance(nb, num.Integral):
            return Zsqrt2(self.a * nb, self.b * nb)
        raise TypeError(f"Multiplication operation is not defined with {type(nb).__name__}")

    def __rmul__(self, nb: int) -> Zsqrt2:
        """Define the right multiplication of int with the Zsqrt2 class."""
        return self.__mul__(nb)

    def __imul__(self, nb: Zsqrt2 | int) -> Zsqrt2:
        """Define in-place multiplication operation for the class."""
        return self.__mul__(nb)

    def __pow__(self, n: int) -> Zsqrt2:
        """Define the power operation for the Zsqrt2 class. Computed using the binomial theorem.

        Exponent must be a positive integer.
        """
        if not isinstance(n, num.Integral):
            raise TypeError(f"Expected power to be of type int, but got {type(n).__name__}.")
        elif n < 0:
            raise ValueError(f"Expected power to be a positive integer, but got {n} < 0.")

        a: int = 0
        b: int = 0
        for k in range(n + 1):
            if k % 2 == 0:
                a += math.comb(n, k) * self.a ** (n - k) * self.b**k * 2 ** (k // 2)
            else:
                b += math.comb(n, k) * self.a ** (n - k) * self.b**k * 2 ** (k // 2)
        return Zsqrt2(a, b)

    def __ipow__(self, nb: int) -> Zsqrt2:
        """Define in-place power."""
        return self.__pow__(nb)

    def __round__(self, precision: Optional[int] = None) -> int | float:
        """Define the round operation.

        Args:
            precision (float, optional): Decimal precision of the rounding. Default = 0.

        Returns:
            (int | float): Ring element rounded to the given precision.
        """
        if precision is None:
            return round(float(self))
        else:
            return round(float(self), precision)

    def __floor__(self) -> int:
        """Define the floor operation."""
        return math.floor(float(self))

    def __ceil__(self) -> int:
        """Define the ceil operation."""
        return math.ceil(float(self))


# LAMBDA = 1 + \u221A2 is used to scale 1D grid problems.
LAMBDA: Zsqrt2 = Zsqrt2(1, 1)

# INVERSE_LAMBDA = -1 + \u221A2 is the inverse of LAMBDA. It is used to scale 1D grid problem.
INVERSE_LAMBDA: Zsqrt2 = -LAMBDA.conjugate()
